Plutarch/List.hs

```diff
@@ -363,3 +370,134 @@ plistEquals =
         )
         (pelimList (\_ _ -> pconstant False) (pconstant True) ylist)
         xlist
+
+preverse :: (PIsListLike l a) => Term s (l a :--> l a)
+preverse = phoistAcyclic $
+  plam $ \xs -> pfoldl # plam (\ys y -> pcons # y # ys) # pnil # xs
+
+pelemAt :: (PIsListLike l a) => Term s (PInteger :--> l a :--> a)
+pelemAt = phoistAcyclic $
+  plam $ \n xs ->
+    pif
+      (n #< 0)
+      perror
+      (pelemAt' # n # xs)
+
+pelemAt' :: (PIsListLike l a) => Term s (PInteger :--> l a :--> a)
+pelemAt' = phoistAcyclic $
+  pfix #$ plam $ \self n xs ->
+    pif
+      (n #== 0)
+      (phead # xs)
+      (self # (n - 1) #$ ptail # xs)
+
+pfind :: (PIsListLike l a) => Term s ((a :--> PBool) :--> l a :--> PMaybe a)
+pfind = phoistAcyclic $
+  pfix #$ plam $ \self f xs ->
+    pelimList
+      ( \y ys ->
+          pif
+            (f # y)
+            (pcon $ PJust y)
+            (self # f # ys)
+      )
+      (pcon PNothing)
+      xs
+
+psort :: POrd a => Term s (PList a :--> PList a)
+psort = ptimSort
+
+-- TODO: decide on default sort
+-- probably by benchmarking them
+
+pmergeSort :: POrd a => Term s (PList a :--> PList a)
+pmergeSort = phoistAcyclic $
+  pfix #$ plam $ \self xs ->
+    pmatch xs $ \case
+      PSNil -> pcon PSNil
+      PSCons x t ->
+        pmatch t $ \case
+          PSNil -> psingleton # x
+          PSCons _ _ -> pmatch (splitList # xs) $
+            \(PPair l r) ->
+              merge # (self # l) # (self # r)
+
+splitList :: Term s (PList a :--> PPair (PList a) (PList a))
+splitList = phoistAcyclic $
+  pfix #$ plam $ \self xs -> pmatch xs $ \case
+    PSNil -> pcon $ PPair pnil pnil
+    PSCons x xs' ->
+      pmatch xs' $ \case
+        PSNil -> pcon $ PPair (psingleton # x) pnil
+        PSCons y ys -> pmatch (self # ys) $
+          \(PPair l r) ->
+            pcon $ PPair (pcons # x # l) (pcons # y # r)
+
+merge :: POrd a => Term s (PList a :--> PList a :--> PList a)
+merge = phoistAcyclic $
+  pfix #$ plam $ \self xs ys ->
+    pmatch xs $ \case
+      PSNil -> ys
+      PSCons x xs' ->
+        pmatch ys $ \case
+          PSNil -> xs
+          PSCons y ys' ->
+            pif
+              (x #<= y)
+              (pcons # x #$ self # xs' # ys)
+              (pcons # y #$ self # xs # ys')
+
+-- most of the tricks from timsort are not implemented here
+-- just the central strategy of finding runs of
+-- consecutive data and then merging them together
+ptimSort :: POrd a => Term s (PList a :--> PList a)
```

Comment:
Why do we need this?

Reply:
I was worried that merge sort might be unnecessarily slow if a lot of the lists it gets called with are nearly sorted. I plan to benchmark the project using both and see what the tradeoff looks like.